src/Search/BreadthFirst.php

```diff
@@ -9,25 +9,41 @@
 class BreadthFirst extends Base
 {
     /**
-     *
+     * @param int $maxDepth
      * @return Vertices
      */
-    public function getVertices()
+    public function getVertices($maxDepth = PHP_INT_MAX)
     {
         $queue = array($this->vertex);
         // to not add vertices twice in array visited
         $mark = array($this->vertex->getId() => true);
         // visited vertices
         $visited = array();
 
+        // keep track of depth
+        $currentDepth = 0;
+        $nodesThisLevel = 1;
+        $nodesNextLevel = 0;
+
         do {
             // get first from queue
             $t = array_shift($queue);
             // save as visited
-            $visited[$t->getId()]= $t;
+            $visited[$t->getId()] = $t;
 
             // get next vertices
-            foreach ($this->getVerticesAdjacent($t)->getMap() as $id => $vertex) {
+            $children = $this->getVerticesAdjacent($t);
+
+            // track depth
+            $nodesNextLevel = $children->count();
+            if (--$nodesThisLevel === 0) {
+                if (++$currentDepth > $maxDepth) goto done;
```

Comment:
Suggested alternative:

```php
if (++$currentDepth > $maxDepth) {
    break;
}
```

Reply:
This will break to foreach($children...) and not done:.

If you're anti goto: even when it's appropriate, I suggest return new Vertices($visited) instead, though I see this as being less DRY.

Reply:
I'm not sure I follow, but the `break` would be within the `do … while` loop only? Either way, I'm not *anti goto* and don't think any of the suggested alternatives really make that much of a difference. The less code, the better; the more structured, the better. This line of code currently violates existing coding standards due to lack of braces (easy fix), while we're at it we *could* replace the `goto` and remove another line. Slightly off-topic, but DRY is not really about lines of code, but more about not making business decisions at different locations which doesn't really apply here.